torchmetrics/functional/classification/auroc.py

```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
     else:
         if mode != DataType.BINARY and num_classes is None:
             raise ValueError('Detected input to `multiclass` but you did not provide `num_classes` argument')
+
+        # If one or more classes has 0 observations, we should exclude them, as its weight will be 0 anyway
+        if average == AverageMethod.WEIGHTED and len(torch.unique(target)) != num_classes:
+            target_bool_mat = torch.zeros((len(target), num_classes), dtype=bool)
+            target_bool_mat[torch.arange(len(target)), target.long()] = 1
+            class_observed = target_bool_mat.sum(axis=0) > 0
+            for c in range(num_classes):
+                if not class_observed[c]:
+                    warnings.warn(f'Class {c} had 0 observations, omitted from AUROC calculation', UserWarning)
+            preds = preds[:, class_observed]
+            target = target_bool_mat[:, class_observed]
+            target = torch.where(target)[1]
+            num_classes = class_observed.sum()
+            if num_classes <= 1:
+                raise ValueError(f'Found {num_classes} non-empty classes in `multiclass` AUROC calculation')
```

Comment:
could we add a test that checks the error gets raised for some invalid input?

Reply:
Done and done!